core/src/test/scala/unit/kafka/controller/ReassignmentsManagerTest.scala

```diff
@@ -0,0 +1,228 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.controller
+
+import kafka.api.LeaderAndIsr
+import kafka.cluster.{Broker, EndPoint}
+import kafka.zk.{KafkaZkClient, TopicPartitionStateZNode}
+import kafka.zk.KafkaZkClient.UpdateLeaderAndIsrResult
+import kafka.zookeeper.SetDataResponse
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.network.ListenerName
+import org.apache.kafka.common.protocol.Errors
+import org.apache.kafka.common.security.auth.SecurityProtocol
+import org.apache.zookeeper.KeeperException.Code
+import org.junit.Assert.{assertEquals, assertTrue}
+import org.junit.{Before, Test}
+import org.mockito.Mockito
+import org.mockito.Mockito.doReturn
+import org.mockito.Mockito.verify
+
+import scala.collection.{Map, Set, mutable}
+
+class ReassignmentsManagerTest {
+  private var controllerContext: ControllerContext = null
+  private var mockZkClient: KafkaZkClient = null
+  private var mockTopicDeletionManager: TopicDeletionManager = null
+  private var mockControllerBrokerRequestBatch: ControllerBrokerRequestBatch = null
+  private var mockReplicaStateMachine: ReplicaStateMachine = null
+  private var mockPartitionStateMachine: PartitionStateMachine = null
+  private var mockBrokerRequestBatch: ControllerBrokerRequestBatch = null
+
+  private final val controllerEpoch = 10
+  private final val zkEpoch = 105
+  private final val topic = "topic"
+  private final val tp = new TopicPartition(topic, 0)
+  private final val mockPartitionReassignmentHandler = new PartitionReassignmentHandler(null)
+
+  private var partitionReassignmentManager: ReassignmentsManager = null
+
+  @Before
+  def setUp(): Unit = {
+    controllerContext = new ControllerContext
+    controllerContext.epoch = controllerEpoch
+    controllerContext.epochZkVersion = zkEpoch
+    mockZkClient = Mockito.mock(classOf[KafkaZkClient])
+    mockTopicDeletionManager = Mockito.mock(classOf[TopicDeletionManager])
+    mockControllerBrokerRequestBatch = Mockito.mock(classOf[ControllerBrokerRequestBatch])
+    mockReplicaStateMachine = Mockito.mock(classOf[ReplicaStateMachine])
+    mockPartitionStateMachine = Mockito.mock(classOf[PartitionStateMachine])
```

Comment:
I'd suggest using `MockPartitionStateMachine` and `MockReplicaStateMachine` to simplify these test cases.

Reply:
Makes sense on the `MockPartitionStateMachine`. With the replica state machine, isn't it more useful to keep it an EasyMock for now since we only use it for the assertion we have in A2 of `testPhaseAOfPartitionReassignment` ?

Reply:
The nice thing that comes from `MockReplicaStateMachine` is validation of the state changes.